src/main/java/com/cloudant/fdblucene/FDBIndexInput.java

```diff
@@ -1,146 +1,130 @@
-    private final long offset;
-    private final long length;
-    private byte[] page;
-    private long pointer;
-    private final int pageSize;
-
-    private final ReadVersionCache readVersionCache;
-
-    FDBIndexInput(final String resourceDescription, final TransactionContext txc, final Subspace subspace,
-            final String name, final long offset, final long length, final int pageSize) {
-        super(resourceDescription);
-        this.txc = txc;
-        this.subspace = subspace;
-        this.name = name;
-        this.offset = offset;
-        this.length = length;
-        page = null;
-        pointer = 0L;
-        this.pageSize = pageSize;
-        this.readVersionCache = new ReadVersionCache();
-    }
-
-    @Override
-    public void close() throws IOException {
-        // empty.
-    }
-
-    @Override
-    public long getFilePointer() {
-        return pointer;
-    }
-
-    @Override
-    public long length() {
-        return length;
+public class FDBIndexInput extends BufferedIndexInput {
+
+  private final TransactionContext txc;
+  private final Subspace subspace;
+  private final String name;
+  private final long off;
+  private final long end;
+  private final int pageSize;
+  private ReadVersionCache readVersionCache;
+
+  public FDBIndexInput(final String resourceDescription, final TransactionContext txc,
+      final Subspace subspace, final String name, final long off, final long length,
+      final int pageSize) {
+    super(resourceDescription, pageSize);
+    this.txc = txc;
+    this.subspace = subspace;
+    this.name = name;
+    this.off = off;
+    this.end = off + length;
+    this.pageSize = pageSize;
+    this.readVersionCache = new ReadVersionCache();
+  }
+
+  @Override
+  protected void readInternal(final byte[] b, final int offset, final int len) throws IOException {
+    long pos = getFilePointer() + off;
+
+    if (pos + len > end) {
+      throw new EOFException("read past EOF: " + this);
     }
 
-    @Override
-    public byte readByte() throws IOException {
-        if (pointer == length) {
-            throw new EOFException("Attempt to read past end of file");
-        }
-
-        loadPageIfNull();
-        byte result = page[FDBUtil.posToOffset(offset + pointer, pageSize)];
-        seek(this.pointer + 1);
-        return result;
+    int readLength = len;
+    int copied = 0;
+    long lastPageNumber = -1L;
+    byte[] page = null;
+    while (readLength > 0) {
+      final long pageNumber = FDBUtil.posToPage(pos, pageSize);
+      if (lastPageNumber != pageNumber) {
+        page = loadPage(pageNumber);
+        lastPageNumber = pageNumber;
+      }
+      final int toCopy = Math.min(page.length - FDBUtil.posToOffset(pos, pageSize), readLength);
+      System.arraycopy(page, FDBUtil.posToOffset(pos, pageSize), b, offset + copied, toCopy);
+      pos += toCopy;
+      copied += toCopy;
+      readLength -= toCopy;
     }
+    assert readLength == 0;
```

Comment:
does this ever happen?

Reply:
it should never happen, that's why it's an assert. :)

Reply:
I modelled this after SimpleFSIndexInput, I even followed variable name conventions, as a way to avoid fence post errors.